exercises/tree_iterators/descr.md

```diff
@@ -0,0 +1,180 @@
+this module is part of OCaml's standard library.
+
+```
+module Seq : sig
+
+  type 'a t = unit -> 'a node
+
+  and +'a node =
+  | Nil
+  | Cons of 'a * 'a t
+
+  val nil : 'a t
+  val cons: 'a -> 'a t -> 'a t
+
+end
+```
+
+This data type is closely related to the algebraic data type of lists.
+Indeed, if instead of `unit -> 'a node` one had written just `'a node`,
+then this data type would have been isomorphic to the type of lists.
+
+The presence of `unit -> ...` indicates that a sequence is in fact a function.
+Calling this function, by applying it to the value `()`, amounts to requesting
+the head of the sequence. This head can be either `Nil`, which means that the
+sequence is empty, or `Cons (x, xs)`, which means that the first element of
+the sequence is `x` and the remaining elements form another sequence `xs`. It
+is worth noting that `xs` is itself a function, so the elements of the
+sequence `xs` need not be explicitly computed until `xs` is applied.
+
+Sequences are closely related to *iterators* in object-oriented languages,
+such as C++ and Java. Yet, sequences are much simpler than iterators, for
+two reasons:
+* they involve no mutable state;
+* they are just as easy to construct and to use as ordinary lists.
+
+The functions `nil` and `cons` are constructor functions. They are provided
+for convenience, but could just as well be defined outside of the module
+`Seq`. Their definitions (not shown) are one line long.
+
+## Building the fringe of a tree (on demand)
+
+We now wish to adapt the code written in Questions 2 and 3
+so as to produce a sequence instead of a list.
+
+**Question 4.**
+Define two mutually recursive functions
+`fringe_seq_with` and
+`fringe_node_with`,
+whose types are as follows:
+```
+  val fringe_seq_with:  'a tree -> 'a Seq.t -> 'a Seq.t
+  val fringe_node_with: 'a tree -> 'a Seq.t -> 'a Seq.node
+```
+such that the sequence `fringe_seq_with t ys`
+begins with the fringe of the tree `t`
+and continues with the sequence `ys`.
+Then, in one line,
+define a function `fringe`
+of type `'a tree -> 'a Seq.t`
+such that `fringe t` is the fringe of the tree `t`.
+
+## Testing the equality of two sequences
+
+**Question 5.**
+Define a function `equal`
+of type `'a Seq.t -> 'a Seq.t -> bool`
+such that `equal xs ys` is `true`
+if and only if
+the sequences `xs` and `ys` are equal.
+The sequences `xs` and `ys` are assumed to be finite.
+Furthermore, we assume that it is permitted to use
+the generic equality function `(=)` to
+compare two elements of type `'a`.
+
+## Testing whether two trees have the same fringe
+
+**Question 6.**
+Define a function `same_fringe`
+of type `'a tree -> 'a tree -> bool`
+such that `same_fringe t1 t2` is `true`
+if and only if the trees `t1` and `t2`
+have the same fringe.
+
+This exercise has real-world applications. For instance, the function
+`same_fringe` can be used to test whether two sets, represented as binary
+search trees, have the same elements.
```

Comment:
Maybe add that one difficulty is to answer `false` as quickly as possible and that the `Seq`-based implementation is both modularly defined and compatible with that requirement.

Reply:
I have added this after Question 6:
Of course, if the trees `t1` and `t2` have distinct fringes,
then the comparison should stop and return `false` as early
as possible.